cli/src/cluster_query.rs

```diff
@@ -844,9 +843,9 @@ pub fn process_catchup(
         );
     }
 
-    let mut previous_rpc_slot = std::u64::MAX;
-    let mut previous_slot_distance = 0;
-    let mut retry_count = 0;
+    let mut previous_rpc_slot = std::i64::MAX;
```

Comment:
The changes here are definitely much better -- what do you think about widening the `i64`s to `i128`s to be totally safe? Then the conversions from u64 will be infallible

Reply:
> The changes here are definitely much better

Thank you :)

> what do you think about widening the `i64`s to `i128`s to be totally safe? Then the conversions from u64 will be infallible

Not sure.

In practice, it is certainly a bug if `i64` is not enough to store a slot index or a slot offset.
So the widening is, in some sense, is just to "hide" the checks.

I do not think we would want to switch to `i128` in other cases.
In particular, in code that is performance sensitive.
So, this would not be a universal solution.

So, I am somewhat against it.
If you insisted, I can make the change, though.

Reply:
I mean that `previous_rpc_slot`, `start_node_slot`, `start_rpc_slot`, `node_slot`, and `rpc_slot` are meant to store `Slot`s, which are `u64`s, so by restricting them into `i64`s, we're introducing a potential bug. We can't safely represent a `Slot` with an `i64`, but we can safely represent it with an `i128`.

If we were in a limited environment, I would certainly let this go, because we have a *long* time until it would ever be a problem. In this case, however, we're just calculating and printing `Slot`s and slot differences, so why not use a type that can always represent a `Slot`?

Reply:
I think our different views on this come down to the "`Slots` are `u64`s" point.  I do not think that any `u64` is a valid slot number.

At `400ms` per slot, we get:
* `32` bits to cover `~54` years.
* `40` bits to cover `~14,000` years.
* `50` bits to cover `~14,000,000` years.

We can probably argue about `2^32` and, maybe, even `2^40`, but I would expect everyone today will agree that a value above `2^50` is not a valid slot index.

From this standpoint, `i64` should be enough to hold a valid absolute slot index or a difference of two indices.  And this is why I am saying that `i128` would just hide an error, without providing a particular benefit.

In fact, Gossip limits Slot values to `1_000_000_000_000_000`:

https://github.com/anza-xyz/agave/blob/be73f30846fb5f651042d3c497e90921ce414685/gossip/src/crds_value.rs#L33

Which is just a bit less than `2^50`:
`log(1_000_000_000_000_000; 2) = 49.8289`.

What do you think,

---

Ideally, it would be better to have a dedicated type that would store `Slot` values.
It would do conversions, checking, provide safer math that does not require explicit checks all over the place, as well as would hold the max value restriction.
Maybe one day :)

Reply:
Ah I see now what you mean, thanks for the explanation. I wasn't aware of the limits elsewhere in the code. By the sounds of things, if we ever do properly type `Slot`, then an `i64`  would safely work. Works for me then!